process_scheduling_algorithms/non_preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,419 @@
+    if (ptr->CT == 0)
+    {
+        CalculateCT(root);
+    }
+    // calculate the total waiting time of all the processes
+    while (ptr != NULL)
+    {
+        ptr->WT = (ptr->TAT - ptr->BT);
+        sum += ptr->WT;
+        ptr = ptr->next;
+    }
+    avg = sum / n;
+    return avg;
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test()
+{
+    // Entered processes
+    // printf("ID Priority Arrival Time Burst Time \n");
+    // printf("1 0 5 1 \n");
+    // printf("2 1 4 2 \n");
+    // printf("3 2 3 3 \n");
+    // printf("4 3 2 4 \n");
+    // printf("5 4 1 5 \n");
+
+    struct node *root = NULL;
+    insert(&root, 1, 0, 5, 1);
+    insert(&root, 2, 1, 4, 2);
+    insert(&root, 3, 2, 3, 3);
+    insert(&root, 4, 3, 2, 4);
+    insert(&root, 5, 4, 1, 5);
+    float avgCT = CalculateCT(&root);
+    float avgTAT = CalculateTAT(&root);
+    float avgWT = CalculateWT(&root);
+    assert(avgCT == 11);
+    assert(avgTAT == 9);
+    assert(avgWT == 6);
+    printf("[+] All tests have successfully passed!\n");
+    // printf("Average Completion Time is : %f \n", CalculateCT(&root));
+    // printf("Average Turn Around Time is : %f \n", CalculateTAT(&root));
+    // printf("Average Waiting Time is : %f \n", CalculateWT(&root));
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main()
+{
+    test();  // run self-test implementations
+
+    // For Menu Driven Program
+    /*
+    struct node *root = NULL;
+    int option = 0, prior, bt, id, at;
+    while (option != 7)
+    {
+        printf("1. Enter Process Details \n");
+        printf("2. Delete Process \n");
+        printf("3. Display Process \n");
+        printf("4. Calculate Avg Completion Time \n");
+        printf("5. Calculate Avg Turn Around Time \n");
+        printf("6. Calculate Avg Waiting Time \n");
+        printf("7. Exit. \n");
+        scanf("%d", &option);
+        switch (option)
+        {
+        case 1:
+            printf("Enter Process Id \n");
+            scanf("%d", &id);
+            printf("Enter Process Priority \n");
+            scanf("%d", &prior);
+            printf("Enter Process Arrival Time \n");
+            scanf("%d", &at);
+            printf("Enter Process Burst Time \n");
+            scanf("%d", &bt);
+            insert(&root, id, at, bt, prior);
+            break;
+        case 2:
+            printf("Enter Process Id \n");
+            scanf("%d", &id);
+            delete (&root, id);
+            break;
+        case 3:
+            showList(root);
+            break;
+        case 4:
+            printf("Average Completion Time is : %f \n", CalculateCT(&root));
+            break;
+        case 5:
+            printf("Average Turn Around Time is : %f \n",
+            CalculateTAT(&root)); break;
+        case 6:
+            printf("Average Waiting Time is : %f \n", CalculateWT(&root));
+            break;
+        default:
+            break;
+        }
+    }
+    */
```

Comment:
Should these comments exists here @Panquesito7?

Reply:
@tjgurwara99 it was to make it menu driven. Should I remove it?

Reply:
I think you should as this adds no real tangible benefit IMO - @Panquesito7 what do you think?

Reply:
Yeah, sounds good. 👍 
I had the idea to keep it in case it's useful for others. It's still going to be saved in the history for those who want to check, anyways.